### PR TITLE
Overhaul Predicate simplifier: Equals, v2

### LINKED ISSUE #2570: Overhaul predicate simplifier
This issue is the first part of addressing the issues raised in #2400. See also: #2566.

We will move `Condition.simplifyPredicate` to its own module and implement a `Predicate.simplify` similar to `TermLike.simplify`. It must not call `simplifyCondition` or `simplifyTermLike`. `Predicate.simplify` will simplify a `Predicate` from the bottom up; at each level, it will iterate until the predicate does not change.

Questions:

- What should be the return type of `Predicate.simplify`? Two reasonable-seeming options are a disjunction of `Condition`s, or a disjunction of `MultiAnd (Predicate _)`. In the former case, the `Condition` will not be simplified. What does the condition simplifier expect?

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes #2570

Completes all cases of `Predicate.simplify`.

I didn't add any tests for `simplifyEquals` because it only calls `Equals.simplify`, which we will need to modify for https://github.com/kframework/kore/issues/2610.

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Simplify/Predicate.hs [hunk 1]
```diff
@@ -180,9 +148,11 @@ simplify sideCondition original =
...
                         >>= simplifyForall sideCondition
+                EqualsF equalsF@(Equals _ _ term _) ->
+                    simplifyEquals sideCondition (termLikeSort term)
+                        =<< traverse simplifyTerm equalsF
                 InF inF ->
                     simplifyIn sideCondition =<< traverse simplifyTerm inF
-                _ -> simplifyPredicateTODO sideCondition predicate & MultiOr.observeAllT
       where
         _ :< predicateF = Recursive.project predicate
         ~avoid = freeVariableNames sideCondition
```
Comment (JKTKops) [inline kore/src/Kore/Simplify/Predicate.hs:158]: Aren't `where` bindings lazy by default? What is this `~` for?

This line wasn't changed, but asking anyway while we're here.

File: kore/src/Kore/Simplify/Predicate.hs [hunk 2]
```diff
@@ -527,6 +497,23 @@ extractFirstAssignment someVariableName predicates =
...
+simplifyEquals ::
+    forall simplifier sort.
+    MonadSimplify simplifier =>
+    SideCondition RewritingVariableName ->
+    Sort ->
+    Equals sort (OrPattern RewritingVariableName) ->
+    simplifier NormalForm
+simplifyEquals sideCondition sort equals =
+    Equals.simplify sideCondition equals'
+        >>= return . MultiOr.map (from @(Condition _))
```
Comment (JKTKops) [inline kore/src/Kore/Simplify/Predicate.hs:509]: Why `m >>= return . f` instead of `fmap f m`?

**Review by JKTKops — APPROVED** [review]
Couple of minor comments, but looks good to me.

File: kore/src/Kore/Simplify/Predicate.hs (hunk 1 above)
Comment (ana-pantilie) [reply kore/src/Kore/Simplify/Predicate.hs:158]: We have `Strict` on by default, maybe this changes the behavior? I don't know for certain, though.

**Pull request merged**